Kmk zarr conversion scripts

kishorikonwar opened the pull request:
Zarr file conversion changes 

(a) test/create_zarr_ss2.py is the python script that changes the SS2 output to zarr format. Note that this script is planned to be added to the sctools repository, but for now, while testing, we are adding it  the docker image "quay.io/humancellatlas/secondary-analysis-python3-scientific:0.1.6_zarr_test".  Eventually, this script will be changed to change the format of the zarr file contents, but for now it tests its ability to create zarr file.

(b)  library/tasks/ZarrUtils.wdl defines the task for the conversion to the zarr format; 
Note that when we are calling the script is it providing the ZipStore option, to create a single file format of the zarr file .  <sample_name>.zip . Note the line 
 output {
    Array[File]  zarr_output_files = glob("*.zip")
  }
in the script is used to capture the output. Although the glob is not required but for DirectoryStore format we might need to it as this format is essentially a hierarchy of folders and files. 

When I tried to use  DirectoryStore option for the zarr library,  the final output in "mint-cromwell job manager" does not contail  final zarr output  <sample_name>.zarr. Note that is is a directory structure (not a single file) but the cromwell gives me error on it" as follows (sample_name is SRR1294925) 
"ln: â€˜SRR1294925.zarrâ€™: hard link not allowed for directory" but the cromwell task runs successfully. This seems to an forlder copying issue from the docker mounted drive to  
"gs://broad-dsde-mint-dev-cromwell-execution/cromwell-executions/SmartSeq2SingleCell/"

(c) pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl is modified to add the zarr file creation task  to the end of the SS2 workflow. 

(d) docker/python3-scientific/Dockerfile is modified to add the python conversion script to create the docker images for the python3-scientific scripts, which is required to run the conversion script"
Note that in this Dockerfile, the python script (in a) is copied to the docker image  for now,  and it is used to  create the docker image "quay.io/humancellatlas/secondary-analysis-python3-scientific:0.1.6_zarr_test" for the tests

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    QC_values = []
+    with open(relevant_qc_files[0], 'r') as f:
+       reader = csv.reader(f)
+       for row in reader:
```
Nice, clear. I'd simplify this a bit, and use the reader to remove the row_name so the rest of the code can be a bit cleaner: 

```python3
with open(relevant_qc_files[0], 'r') as f:
    reader = csv.reader(f)
    _ = next(reader)  # first line is a header, which we don't use. 
    QC_values = [row for row in scv.reader(f)]
```

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    datagroup = root['QCs']
+
+    # Metrics
+    """ The first value is the row name so we remove it"""
```
I'd use an in-line comment with `#` since the comment doesn't need multiple lines.

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    
+    metrics_array=datagroup.create_dataset(
+        "QCs",
+        shape=(len(QC_values[0][1:]), ),
```
Combining my comments above, this can turn into `shape=(len(QC_values), )`. Similar improvements are possible for Metrics, Class, and Values.

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+        data = QC_values[0][1:]
+     )
+
+    # Class
```
What is Class?

kishorikonwar replied:
It was a category Jishu assigns in the <sample_name>_QCs.csv file but I removed it for now..as is it s confusing and the information is captured in qc_metric

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+        Raises:
+            FileNotFoundError
+    """
+    #expression_path = glob.glob(os.path.join(input_path, _FILES['genes_results'][1]))[0]
```
Did you mean to remove this? If no, and it's meant to be instructive, add a comment above about how it should be interpreted.

ambrosejcarr commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    args = parser.parse_args()
+
+    create_zarr_files(args.sample_id, args.qc_analysis_output_files_string, args.rsem_genes_results, args.output_zarr_path, args.format)
+    #add_cell_metadata(root, args.analysis_output_path)
```
remove commented lines.

kbergin commented:
@chrisfarnham

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+  # converts some of the outputs of Smart Seq 2 pipeline into a zarr file
+
+  File rsem_gene_results # the gene count file
+  Array[File] smartseq_qc_files   # location of the smart seq 2 output
```
I think we can ask Jishu. I agree with you because this was making the task difficult

kishorikonwar commented on docker/python3-scientific/Dockerfile:
```diff
@@ -8,13 +8,22 @@ RUN apt update && apt install -y \
     samtools
 
 RUN pip3 install \
```
Just to upkeep the container. We would not need this if we created a separate docker name. As of now we simply added the script to the docker image "quay.io/humancellatlas/secondary-analysis-python3-scientific"

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+task SmartSeq2ZarrConversion {
+  # converts some of the outputs of Smart Seq 2 pipeline into a zarr file
+
+  File rsem_gene_results # the gene count file
```
Addressed!

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+task SmartSeq2ZarrConversion {
+  # converts some of the outputs of Smart Seq 2 pipeline into a zarr file
```
Addressed.

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+  # converts some of the outputs of Smart Seq 2 pipeline into a zarr file
+
+  File rsem_gene_results # the gene count file
+  Array[File] smartseq_qc_files   # location of the smart seq 2 output
```
Updated the comment to convey the information accurately.

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+
+  File rsem_gene_results # the gene count file
+  Array[File] smartseq_qc_files   # location of the smart seq 2 output
+  String sample_name   # id of the sample
```
Addressed.

kishorikonwar commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    
+    # split the files string by , 
+    # crate an array of the file names containing the suffix _QCs.csv 
+    relevant_qc_files =  [  x.strip() for x in input_files_string.split(',')  if re.search(r'_QCs.csv$', x) ]
```
changed

kishorikonwar commented on docker/python3-scientific/Dockerfile:
```diff
@@ -8,13 +8,22 @@ RUN apt update && apt install -y \
     HTSeq==0.9.0 \
     matplotlib==2.1.0 \
-    numpy==1.12.0 \
+    numpy==1.15.2 \
```
It was not necessary to make it work. So I reverted back to 1.12.0

kishorikonwar commented on docker/python3-scientific/Dockerfile:
```diff
@@ -8,13 +8,22 @@ RUN apt update && apt install -y \
+
+WORKDIR /tools
+
+COPY create_zarr_ss2.py .
```
We are thinking of a adding it to "quay.io/humancellatlas/secondary-analysis-python3-scientific" I will consult rex on this

kishorikonwar commented:
@mckinsel @ambrosejcarr @rexwangcc 
Hi Guys,
I just pushed the changes. Here are a few things:
1. Addressed the concerns and suggestions of the code reviewers
2. Please see the comment-wise response in the PR
3. The zarr conversion in the flattened form works now, i.e., the zarr folder structure is flattened to gs by replacing "/" with "!"

kishorikonwar commented:
@rexwangcc @mckinsel  @ambrosejcarr 
Hi Guys,

I addressed Rex's comments in the new push

kishorikonwar commented on docker/python3-scientific/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM python:3.6.2
+FROM quay.io/humancellatlas/secondary-analysis-python3-scientific:0.1.6
```
We are taking approach 2)

rexwangcc commented:
FYI This PR is passing on Mintegration test: https://job-manager.caas-prod.broadinstitute.org/jobs/285b5d53-af4c-420c-b81e-7abd16175594 with another PR in secondary-analysis repo:  https://github.com/HumanCellAtlas/secondary-analysis/pull/359

samanehsan commented:
Hi @kishorikonwar, please include the `ZarrUtils.wdl` WDL here so that the Jenkins SS2 Tests uses it as well: https://github.com/HumanCellAtlas/skylab/blob/master/test/smartseq2_single_sample/pr/dependencies.json

kishorikonwar commented:
@mckinsel @ambrosejcarr @rexwangcc  @chrisfarnham 
We make a change to exclude the python script from the output. Note that previously our glob expression had the form "*zarr*" but our script has the substring "zarr". As a result, the workflow output included the script. In the last commit we moved the the output to a temp folder and applied the right zarr expression, to include the temp folder. This solves the problem. Please see the Mintegrat test results in 

https://job-manager.caas-prod.broadinstitute.org/jobs/c556b08c-f00c-44b0-abf9-a45b8c0c0f7d
and the resulting zarr output in the [bucket](https://console.cloud.google.com/storage/browser/broad-dsde-mint-test-cromwell-execution/caas-cromwell-executions/AdapterSmartSeq2SingleCell/c556b08c-f00c-44b0-abf9-a45b8c0c0f7d/call-analysis/ss2.SmartSeq2SingleCell/1c2a301d-af2f-4bd8-97c4-5f47f4539269/call-SmartSeq2ZarrConversion/glob-1499f05c828563a73cd6a8ebef31c37a/)

kishorikonwar commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,34 @@
+
+    set -e
+
+    cp /tools/create_zarr_ss2.py ./
```
Changed.

kishorikonwar commented on test/create_zarr_ss2.py:
```diff
@@ -0,0 +1,194 @@
+    root = zarr.group(store, overwrite=True)
+
+    # add some readme for the user
+    root.attrs['README'] = "The schema adopted in this zarr store may undergo changes in future"
```
Fixed.

kishorikonwar commented on pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl:
```diff
@@ -132,5 +140,8 @@ workflow SmartSeq2SingleCell {
     File rsem_isoform_results = RSEMExpression.rsem_isoform
+
+    # zarr
+    Array [File] zarr_output_files = SmartSeq2ZarrConversion.zarr_output_files
```
Fixed.

mckinsel commented on library/tasks/ZarrUtils.wdl:
```diff
@@ -0,0 +1,54 @@
+
+    mkdir zarrout
+    #get all the files in the zarr folder in  a list
+    a=`find "${sample_name}.zarr"  -type f`
```
In general, it's not a great idea to iterate over the output of `find`. See [here](https://mywiki.wooledge.org/BashPitfalls#for_f_in_.24.28ls_.2A.mp3.29) for a discussion.

rexwangcc replied:
@mckinsel So do you have any preferences on this? (I might use some Python code here to do the iteration rather than bash `find`.)

chrisfarnham replied:
I am recommending that we merge this tested and functional feature into the greater system today. We will open a new ticket that captures this requested improvement for later releases. 

We have a solution to this comment written but it is untested. Merging the PR as it is now clears the path for other blocking PRs. Feel free to reach out to me directly for discussion.

kbergin replied:
+1 to avoiding scope creep, please someone make said ticket in secondary-analysis and leave it in 'new issues' for me

mckinsel replied:
Yeah for sure, flag for later.

rexwangcc commented:
@jishuxu @ambrosejcarr @mckinsel Could any of the CBs gives a 👍 to this PR so we can merge it ASAP?

rexwangcc commented:
retest this please

samanehsan commented:
🎉 🎉 🎉 🎉 🎉
